src/sources/geniuscom.cpp

```diff
@@ -71,57 +75,112 @@ std::vector<LyricDataRaw> GeniusComSource::search(const LyricSearchParams& param
     }
 
-    LOG_INFO("Page %s retrieved", url.c_str());
-    std::string lyric_text;
-    pugi::xml_document doc;
-    load_html_document(content.c_str(), doc);
+    std::vector<LyricDataRaw> song_metadata;
+    {   // Parser gets its own scope
+        cJSON* json = cJSON_Parse(content.c_str());
+        if (json == nullptr) {
+            LOG_WARN("Failed to parse genius.com search result %s", content.c_str());
+            cJSON_Delete(json);
+            return {};
+        }
 
-    const char* xpath_queries[] = { "//div[@class='lyrics']", "//div[contains(@class, 'Lyrics__Container')]" };
-    for(const char* query_str : xpath_queries)
-    {
-        pugi::xpath_query query_lyricdivs(query_str);
-        pugi::xpath_node_set lyricdivs = query_lyricdivs.evaluate_node_set(doc);
+        const cJSON* search_response = cJSON_GetObjectItem(json, "response");
+        const cJSON* search_hits = cJSON_GetObjectItem(search_response, "hits");
 
-        if(!lyricdivs.empty())
-        {
-            for(const pugi::xpath_node& node : lyricdivs)
-            {
-                add_all_text_to_string(lyric_text, node.node());
-
-                // A div is a block element, which means that by definition
-                // it effectively includes a trailing line-break.
-                // We won't get that line-break by parsing the HTML text content,
-                // so add it here manually.
-                lyric_text += "\r\n";
+        if (!cJSON_IsArray(search_hits)) {
+            LOG_INFO("Failed to download genius.com page: Invalid search payload");
+            cJSON_Delete(json);
+            return {};
+        }
+
+        int results = 0;
+        const cJSON* search_hit = nullptr;
+        cJSON_ArrayForEach(search_hit, search_hits) {
+            results++;
+            if (results > RESULT_LIMIT) {
+                break;
             }
-            break;
+
+			const cJSON* search_result = cJSON_GetObjectItem(search_hit, "result");
+			const cJSON* search_path = cJSON_GetObjectItem(search_result, "api_path");
+			const cJSON* search_title = cJSON_GetObjectItem(search_result, "title");
+			const cJSON* search_artist = cJSON_GetObjectItem(search_result, "artist_names");
+
+			LyricDataRaw result = {};
+			result.source_id = id();
+			result.lookup_id = search_path->valuestring;
+            result.source_path = search_path->valuestring;
+            result.title = search_title->valuestring;
+            result.artist = search_artist->valuestring;
+
+			song_metadata.push_back(result);
         }
+
+        cJSON_Delete(json);
     }
 
-    if(lyric_text.empty())
+    return song_metadata;
+}
+
+bool GeniusComSource::lookup(LyricDataRaw& data, abort_callback& abort)
+{
+    auto request = http_client::get()->create_request("GET");
+    request->add_header(API_KEY_HEADER);
+
+    std::string url = "https://api.genius.com" + std::string(data.lookup_id) + "?text_format=plain";
+
+    LOG_INFO("Page %s retrieved", url.c_str());

```

Comment:
False, we've not retrieved anything at this point. That happens further down and we already log about it (on line 148)